File: frontend/src/api/interface/website.ts
import { CommonModel, ReqPage } from '.';

export namespace Website {
    export interface Website extends CommonModel {
        primaryDomain: string;
        type: string;
        alias: string;
        remark: string;
        domains: string[];
        appType: string;
        appInstallId?: number;
        webSiteGroupId: number;
        otherDomains: string;
        defaultServer: boolean;
        protocol: string;
        autoRenew: boolean;
        appinstall?: NewAppInstall;
        webSiteSSL: SSL;
        runtimeID: number;
        rewrite: string;
        user: string;
        group: string;
        IPV6: boolean;
        accessLog?: boolean;
        errorLog?: boolean;
        childSites?: string[];
        dbID: number;
        dbType: string;
        favorite: boolean;
    }

    export interface WebsiteDTO extends Website {
        errorLogPath: string;
        accessLogPath: string;
        sitePath: string;
        appName: string;
        runtimeName: string;
        runtimeType: string;
        openBaseDir: boolean;
    }
    export interface WebsiteRes extends CommonModel {
        protocol: string;
        primaryDomain: string;
        type: string;
        alias: string;
        remark: string;
        status: string;
        expireDate: string;
        sitePath: string;
        appName: string;
        runtimeName: string;
        sslExpireDate: Date;
    }

    export interface NewAppInstall {
        name: string;
        appDetailId: number;
        params: any;
    }

    export interface WebSiteSearch extends ReqPage {
        name: string;
        orderBy: string;
        order: string;
        websiteGroupId: number;
    }

    export interface WebSiteDel {
        id: number;
        deleteApp: boolean;
        deleteBackup: boolean;
        forceDelete: boolean;
    }

    export interface WebSiteCreateReq {
        type: string;
        alias: string;
        remark: string;
        appType: string;
        appInstallId: number;
        webSiteGroupId: number;
        proxy: string;
        proxyType: string;
        ftpUser: string;
        ftpPassword: string;
        taskID: string;
        SSLID?: number;
        enableSSL: boolean;
        createDB?: boolean;
        dbName?: string;
        dbPassword?: string;
        dbFormat?: string;
        dbUser?: string;
        dbHost?: string;
        domains: SubDomain[];
    }

    export interface WebSiteUpdateReq {
        id: number;
        primaryDomain: string;
        remark: string;
        webSiteGroupId: number;
        expireDate?: string;
        IPV6: boolean;
        favorite: boolean;
    }

    export interface WebSiteOp {
        id: number;
        operate: string;
    }

    export interface WebSiteOpLog {
        id: number;
        operate: string;
        logType: string;
        page?: number;
        pageSize?: number;
    }

    export interface OptionReq {
        types?: string[];
    }

    export interface WebSiteLog {
        enable: boolean;
        content: string;
        end: boolean;
        path: string;
    }

    export interface Domain {
        websiteId: number;
        port: number;
        id: number;
        domain: string;
        ssl: boolean;
    }

    export interface DomainCreate {
        websiteID: number;
        domains: SubDomain[];
    }

    export interface DomainUpdate {
        id: number;
        ssl: boolean;
    }

    interface SubDomain {
        domain: string;
        port: number;
        ssl: boolean;
    }

    export interface DomainDelete {
        id: number;
    }

    export interface NginxConfigReq {
        operate: string;
        websiteId: number;
        scope: string;
        params?: any;
    }

    export interface NginxScopeReq {
        websiteId: number;
        scope: string;
    }

    export interface NginxParam {
        name: string;
        params: string[];
    }

    export interface NginxScopeConfig {
        enable: boolean;
        params: NginxParam[];
    }

    export interface DnsAccount extends CommonModel {
        name: string;
        type: string;
        authorization: Object;
    }

    export interface DnsAccountCreate {
        name: string;
        type: string;
        authorization: Object;
    }

    export interface DnsAccountUpdate {
        id: number;
        name: string;
        type: string;
        authorization: Object;
    }

    export interface SSL extends CommonModel {
        primaryDomain: string;
        privateKey: string;
        pem: string;
        otherDomains: string;
        certURL: string;
        type: string;
        issuerName: string;
        expireDate: string;
        startDate: string;
        provider: string;
        websites?: Website.Website[];
        autoRenew: boolean;
        acmeAccountId: number;
        status: string;
        domains: string;
        description: string;
        dnsAccountId?: number;
        pushDir: boolean;
        dir: string;
        keyType: string;
        nameserver1: string;
        nameserver2: string;
        disableCNAME: boolean;
        skipDNS: boolean;
        execShell: boolean;
        shell: string;
    }

    export interface SSLDTO extends SSL {
        logPath: string;
    }

    export interface SSLCreate {
        primaryDomain: string;
        otherDomains: string;
        provider: string;
        acmeAccountId: number;
        dnsAccountId: number;
        id?: number;
        description: string;
    }

    export interface SSLApply {
        websiteId: number;
        SSLId: number;
    }

    export interface SSLRenew {
        SSLId: number;
    }

    export interface SSLUpdate {
        id: number;
        autoRenew: boolean;
        description: string;
        primaryDomain: string;
        otherDomains: string;
        acmeAccountId: number;
        provider: string;
        dnsAccountId?: number;
        keyType: string;
        pushDir: boolean;
        dir: string;
    }

    export interface AcmeAccount extends CommonModel {
        email: string;
        url: string;
        type: string;
        useProxy: boolean;
    }

    export interface AcmeAccountCreate {
        email: string;
        useProxy: boolean;
    }

    export interface AcmeAccountUpdate {
        id: number;
        useProxy: boolean;
    }

    export interface DNSResolveReq {
        acmeAccountId: number;
        websiteSSLId: number;
    }

    export interface DNSResolve {
        resolve: string;
        value: string;
        domain: string;
        err: string;
    }

    export interface SSLReq {
        name?: string;
        acmeAccountID?: string;
    }

    export interface HTTPSReq {
        websiteId: number;
        enable: boolean;
        websiteSSLId?: number;
        type: string;
        certificate?: string;
        privateKey?: string;
        httpConfig: string;
        SSLProtocol: string[];
        algorithm: string;
        http3: boolean;
    }

    export interface HTTPSConfig {
        enable: boolean;
        SSL: SSL;
        httpConfig: string;
        SSLProtocol: string[];
        algorithm: string;
        hsts: boolean;
        hstsIncludeSubDomains: boolean;
        httpsPort?: string;
        http3: boolean;
    }

    export interface CheckReq {
        installIds?: number[];
    }

    export interface CheckRes {
        name: string;
        status: string;
        version: string;
        appName: string;
    }

    export interface DelReq {
        id: number;
    }

    export interface NginxUpdate {
        id: number;
        content: string;
    }

    export interface DefaultServerUpdate {
        id: number;
    }
    export interface RewriteReq {
        websiteID: number;
        name: string;
    }

    export interface RewriteRes {
        content: string;
    }

    export interface RewriteUpdate {
        websiteID: number;
        name: string;
        content: string;
    }

    export interface CustomRewirte {
        operate: string;
        name: string;
        content: string;
    }

    export interface DirUpdate {
        id: number;
        siteDir: string;
    }

    export interface DirPermissionUpdate {
        id: number;
        user: string;
        group: string;
    }

    export interface ProxyReq {
        id: number;
    }

    export interface ProxyConfig {
        id: number;
        operate: string;
        enable: boolean;
        cache: boolean;
        cacheTime: number;
        cacheUnit: string;
        serverCacheTime: number;
        serverCacheUnit: string;
        name: string;
        modifier: string;
        match: string;
        proxyPass: string;
        proxyHost: string;
        filePath?: string;
        replaces?: ProxReplace;
        content?: string;
        proxyAddress?: string;
        proxyProtocol?: string;
        sni?: boolean;
        proxySSLName: string;
    }

    export interface ProxReplace {
        [key: string]: string;
    }

    export interface ProxyFileUpdate {
        websiteID: number;
        name: string;
        content: string;
    }

    export interface AuthReq {
        websiteID: number;
    }

    export interface NginxAuth {
        username: string;
        remark: string;
    }

    export interface AuthConfig {
        enable: boolean;
        items: NginxAuth[];
    }

    export interface NginxAuthConfig {
        websiteID: number;
        operate: string;
        username: string;
        password: string;
        remark: string;
        scope: string;
        path?: '';
        name?: '';
    }

    export interface NginxPathAuthConfig {
        websiteID: number;
        operate: string;
        path: string;
        username: string;
        password: string;
        name: string;
    }

    export interface LeechConfig {
        enable: boolean;
        cache: boolean;
        cacheTime: number;
        cacheUint: string;
        extends: string;
        return: string;
        serverNames: string[];
        noneRef: boolean;
        logEnable: boolean;
        blocked: boolean;
        websiteID?: number;
    }

    export interface LeechReq {
        websiteID: number;
    }

    export interface WebsiteReq {
        websiteID: number;
    }

    export interface RedirectConfig {
        operate: string;
        websiteID: number;
        domains?: string[];
        enable: boolean;
        name: string;
        keepPath: boolean;
        type: string;
        redirect: string;
        path?: string;
        target: string;
        redirectRoot?: boolean;
        filePath?: string;
        content?: string;
    }

    export interface RedirectFileUpdate {
        websiteID: number;
        name: string;
        content: string;
    }

    export interface PHPVersionChange {
        websiteID: number;
        runtimeID: number;
    }

    export interface DirConfig {
        dirs: string[];
        user: string;
        userGroup: string;
        msg: string;
    }

    export interface SSLUpload {
        privateKey: string;
        certificate: string;
        privateKeyPath: string;
        certificatePath: string;
        type: string;
        sslID: number;
    }

    export interface SSLObtain {
        ID: number;
    }

    export interface CA extends CommonModel {
        name: string;
        csr: string;
        privateKey: string;
        keyType: string;
    }

    export interface CACreate {
        name: string;
        commonName: string;
        country: string;
        organization: string;
        organizationUint: string;
        keyType: string;
        province: string;
        city: string;
    }

    export interface CADTO extends CA {
        commonName: string;
        country: string;
        organization: string;
        organizationUint: string;
        province: string;
        city: string;
    }

    export interface SSLObtainByCA {
        id: number;
        domains: string;
        keyType: string;
        time: number;
        unit: string;
        pushDir: boolean;
        dir: string;
        description: string;
    }

    export interface RenewSSLByCA {
        SSLID: number;
    }

    export interface SSLDownload {
        id: number;
    }

    export interface WebsiteHtml {
        content: string;
    }
    export interface WebsiteHtmlUpdate {
        type: string;
        content: string;
    }

    export interface NginxUpstream {
        name: string;
        algorithm: string;
        servers: NginxUpstreamServer[];
        content?: string;
        websiteID?: number;
    }

    export interface NginxUpstreamFile {
        name: string;
        content: string;
        websiteID: number;
    }

    export interface LoadBalanceReq {
        websiteID: number;
        name: string;
        algorithm: string;
        servers: NginxUpstreamServer[];
    }

    interface NginxUpstreamServer {
        server: string;
        weight: number;
        failTimeout: string;
        maxFails: number;
        maxConns: number;
        flag: string;
    }

    export interface LoadBalanceDel {
        websiteID: number;
        name: string;
    }

    export interface WebsiteLBUpdateFile {
        websiteID: number;
        name: string;
        content: string;
    }

    export interface WebsiteCacheConfig {
        open: boolean;
        cacheLimit: number;
        cacheLimitUnit: string;
        shareCache: number;
        shareCacheUnit: string;
        cacheExpire: number;
        cacheExpireUnit: string;
    }

    export interface WebsiteRealIPConfig {
        open: boolean;
        ipFrom: string;
        ipHeader: string;
        ipOther: string;
    }

    export interface WebsiteResource {
        name: string;
        type: string;
        resourceID: number;
        detail: any;
    }

    export interface WebsiteDatabase {
        type: string;
        databaseID: number;
        websiteID: number;
        from: string;
        databaseName: number;
    }

    export interface ChangeDatabase {
        websiteID: number;
        databaseID: number;
        databaseType: string;
    }

    export interface CrossSiteAccessOp {
        websiteID: number;
        operation: string;
    }

    export interface ExecComposer {
        websiteID: number;
        command: string;
        extCommand?: string;
        mirror: string;
        dir: string;
        user: string;
        taskID: string;
    }
}

File: agent/app/dto/request/website.go
package request

import (
	"github.com/1Panel-dev/1Panel/agent/app/dto"
)

type WebsiteSearch struct {
	dto.PageInfo
	Name           string `json:"name"`
	OrderBy        string `json:"orderBy" validate:"required,oneof=primary_domain type status createdAt expire_date created_at favorite"`
	Order          string `json:"order" validate:"required,oneof=null ascending descending"`
	WebsiteGroupID uint   `json:"websiteGroupId"`
	Type           string `json:"type"`
}

type WebsiteCreate struct {
	Type           string `json:"type" validate:"required"`
	Alias          string `json:"alias" validate:"required"`
	Remark         string `json:"remark"`
	Proxy          string `json:"proxy"`
	WebsiteGroupID uint   `json:"webSiteGroupID" validate:"required"`
	IPV6           bool   `json:"IPV6"`

	Domains []WebsiteDomain `json:"domains"`

	AppType      string        `json:"appType" validate:"oneof=new installed"`
	AppInstall   NewAppInstall `json:"appInstall"`
	AppID        uint          `json:"appID"`
	AppInstallID uint          `json:"appInstallID"`

	RuntimeID       uint   `json:"runtimeID"`
	TaskID          string `json:"taskID"`
	ParentWebsiteID uint   `json:"parentWebsiteID"`

	SiteDir string `json:"siteDir"`

	RuntimeConfig
	FtpConfig
	DataBaseConfig
	SSLConfig
}

type WebsiteOptionReq struct {
	Types []string `json:"types"`
}

type RuntimeConfig struct {
	ProxyType string `json:"proxyType"`
	Port      int    `json:"port"`
}

type FtpConfig struct {
	FtpUser     string `json:"ftpUser"`
	FtpPassword string `json:"ftpPassword"`
}

type DataBaseConfig struct {
	CreateDb   bool   `json:"createDb"`
	DbName     string `json:"dbName"`
	DbUser     string `json:"dbUser"`
	DbPassword string `json:"dbPassword"`
	DbHost     string `json:"dbHost"`
	DBFormat   string `json:"dbFormat"`
}

type SSLConfig struct {
	EnableSSL    bool `json:"enableSSL"`
	WebsiteSSLID uint `json:"websiteSSLID"`
}

type NewAppInstall struct {
	Name        string                 `json:"name"`
	AppDetailId uint                   `json:"appDetailID"`
	Params      map[string]interface{} `json:"params"`

	AppContainerConfig
}

type WebsiteInstallCheckReq struct {
	InstallIds []uint `json:"InstallIds"`
}

type WebsiteUpdate struct {
	ID             uint   `json:"id" validate:"required"`
	PrimaryDomain  string `json:"primaryDomain" validate:"required"`
	Remark         string `json:"remark"`
	WebsiteGroupID uint   `json:"webSiteGroupID"`
	ExpireDate     string `json:"expireDate"`
	IPV6           bool   `json:"IPV6"`
	Favorite       bool   `json:"favorite"`
}

type WebsiteDelete struct {
	ID           uint `json:"id" validate:"required"`
	DeleteApp    bool `json:"deleteApp"`
	DeleteBackup bool `json:"deleteBackup"`
	ForceDelete  bool `json:"forceDelete"`
}

type WebsiteOp struct {
	ID      uint   `json:"id" validate:"required"`
	Operate string `json:"operate"`
}

type WebsiteRedirectUpdate struct {
	WebsiteID uint   `json:"websiteId" validate:"required"`
	Key       string `json:"key" validate:"required"`
	Enable    bool   `json:"enable"`
}

type WebsiteRecover struct {
	WebsiteName string `json:"websiteName" validate:"required"`
	Type        string `json:"type" validate:"required"`
	BackupName  string `json:"backupName" validate:"required"`
}

type WebsiteRecoverByFile struct {
	WebsiteName string `json:"websiteName" validate:"required"`
	Type        string `json:"type" validate:"required"`
	FileDir     string `json:"fileDir" validate:"required"`
	FileName    string `json:"fileName" validate:"required"`
}

type WebsiteGroupCreate struct {
	Name string `json:"name" validate:"required"`
}

type WebsiteGroupUpdate struct {
	ID      uint   `json:"id" validate:"required"`
	Name    string `json:"name"`
	Default bool   `json:"default"`
}

type WebsiteDomainCreate struct {
	WebsiteID uint            `json:"websiteID" validate:"required"`
	Domains   []WebsiteDomain `json:"domains" validate:"required"`
}

type WebsiteDomainUpdate struct {
	ID  uint `json:"id" validate:"required"`
	SSL bool `json:"ssl"`
}

type WebsiteDomain struct {
	Domain string `json:"domain" validate:"required"`
	Port   int    `json:"port"`
	SSL    bool   `json:"ssl"`
}

type WebsiteDomainDelete struct {
	ID uint `json:"id" validate:"required"`
}

type WebsiteHTTPSOp struct {
	WebsiteID       uint     `json:"websiteId" validate:"required"`
	Enable          bool     `json:"enable"`
	WebsiteSSLID    uint     `json:"websiteSSLId"`
	Type            string   `json:"type"  validate:"oneof=existed auto manual"`
	PrivateKey      string   `json:"privateKey"`
	Certificate     string   `json:"certificate"`
	PrivateKeyPath  string   `json:"privateKeyPath"`
	CertificatePath string   `json:"certificatePath"`
	ImportType      string   `json:"importType"`
	HttpConfig      string   `json:"httpConfig"  validate:"oneof=HTTPSOnly HTTPAlso HTTPToHTTPS"`
	SSLProtocol     []string `json:"SSLProtocol"`
	Algorithm       string   `json:"algorithm"`
	Hsts            bool     `json:"hsts"`
	HstsIncludeSubDomains            bool     `json:"hstsIncludeSubDomains"`
	HttpsPorts      []int    `json:"httpsPorts"`
	Http3           bool     `json:"http3"`
}

type WebsiteNginxUpdate struct {
	ID      uint   `json:"id" validate:"required"`
	Content string `json:"content" validate:"required"`
}

type WebsiteLogReq struct {
	ID       uint   `json:"id" validate:"required"`
	Operate  string `json:"operate" validate:"required"`
	LogType  string `json:"logType" validate:"required"`
	Page     int    `json:"page"`
	PageSize int    `json:"pageSize"`
}

type WebsiteDefaultUpdate struct {
	ID uint `json:"id"`
}

type WebsitePHPVersionReq struct {
	WebsiteID uint `json:"websiteID" validate:"required"`
	RuntimeID uint `json:"runtimeID"`
}

type WebsiteUpdateDir struct {
	ID      uint   `json:"id" validate:"required"`
	SiteDir string `json:"siteDir" validate:"required"`
}

type WebsiteUpdateDirPermission struct {
	ID    uint   `json:"id" validate:"required"`
	User  string `json:"user" validate:"required"`
	Group string `json:"group" validate:"required"`
}

type WebsiteProxyConfig struct {
	ID              uint              `json:"id" validate:"required"`
	Operate         string            `json:"operate" validate:"required"`
	Enable          bool              `json:"enable" `
	Cache           bool              `json:"cache" `
	CacheTime       int               `json:"cacheTime"`
	CacheUnit       string            `json:"cacheUnit"`
	ServerCacheTime int               `json:"serverCacheTime"`
	ServerCacheUnit string            `json:"serverCacheUnit"`
	Name            string            `json:"name" validate:"required"`
	Modifier        string            `json:"modifier"`
	Match           string            `json:"match" validate:"required"`
	ProxyPass       string            `json:"proxyPass" validate:"required"`
	ProxyHost       string            `json:"proxyHost" validate:"required"`
	Content         string            `json:"content"`
	FilePath        string            `json:"filePath"`
	Replaces        map[string]string `json:"replaces"`
	SNI             bool              `json:"sni"`
	ProxySSLName    string            `json:"proxySSLName"`
}

type WebsiteProxyReq struct {
	ID uint `json:"id" validate:"required"`
}

type WebsiteRedirectReq struct {
	WebsiteID uint `json:"websiteId" validate:"required"`
}

type WebsiteCommonReq struct {
	ID uint `json:"id" validate:"required"`
}

type WafWebsite struct {
	Key     string   `json:"key"`
	Domains []string `json:"domains"`
	Host    []string `json:"host"`
}

type WebsiteHtmlReq struct {
	Type string `json:"type" validate:"required"`
}

type WebsiteHtmlUpdate struct {
	Type    string `json:"type" validate:"required"`
	Content string `json:"content" validate:"required"`
}

type WebsiteLBCreate struct {
	WebsiteID uint                      `json:"websiteID" validate:"required"`
	Name      string                    `json:"name" validate:"required"`
	Algorithm string                    `json:"algorithm"`
	Servers   []dto.NginxUpstreamServer `json:"servers"`
}

type WebsiteLBUpdate struct {
	WebsiteID uint                      `json:"websiteID" validate:"required"`
	Name      string                    `json:"name" validate:"required"`
	Algorithm string                    `json:"algorithm"`
	Servers   []dto.NginxUpstreamServer `json:"servers"`
}

type WebsiteLBDelete struct {
	WebsiteID uint   `json:"websiteID" validate:"required"`
	Name      string `json:"name" validate:"required"`
}

type WebsiteLBUpdateFile struct {
	WebsiteID uint   `json:"websiteID" validate:"required"`
	Name      string `json:"name" validate:"required"`
	Content   string `json:"content" validate:"required"`
}

type WebsiteRealIP struct {
	WebsiteID uint   `json:"websiteID" validate:"required"`
	Open      bool   `json:"open"`
	IPFrom    string `json:"ipFrom"`
	IPHeader  string `json:"ipHeader"`
	IPOther   string `json:"ipOther"`
}

type ChangeDatabase struct {
	WebsiteID    uint   `json:"websiteID" validate:"required"`
	DatabaseID   uint   `json:"databaseID" `
	DatabaseType string `json:"databaseType" `
}

type WebsiteProxyDel struct {
	ID   uint   `json:"id" validate:"required"`
	Name string `json:"name" validate:"required"`
}

type CrossSiteAccessOp struct {
	WebsiteID uint   `json:"websiteID" validate:"required"`
	Operation string `json:"operation" validate:"required,oneof=Enable Disable"`
}
